### PR TITLE
[WIP] Speed up AT+BLEKEYBOARDCODE

### PR DESCRIPTION
**Pull request by Gigahawk**
Using `command()` and `command_check_OK()` locks up the processor for a significant amount of time.

This is problematic for use cases where we need to hand back control to the caller as soon as possible, such as scanning a keyboard matrix.

QMK handles this by storing SDEP packets in a queue, then periodically shifting them out every time the main loop runs.

This PR aims to implement something similar, with a `send_keyboard_code()` function to push the requisite SDEP packets into a FIFO queue, then a `pop_keyboard_code_queue()` function to be called by a matrix scanning loop as required.

There's a proof of concept implementation of this on my KMK fork here: https://github.com/Gigahawk/kmk_firmware/blob/bluefruit_spi/kmk/drivers/bluefruitspi.py, which should be testable using the following main.py:
```python
import board
from digitalio import DigitalInOut, Direction, Pull

from kmk.hid import HIDModes
from kmk.keys import KC
from kmk.kmk_keyboard import KMKKeyboard
from kmk.matrix import DiodeOrientation

DEBUG_ENABLE = True

keyboard = KMKKeyboard()

keyboard.debug_enabled = False

keyboard.col_pins = (board.A2, board.A3)
keyboard.row_pins = (board.D5, board.D6)
keyboard.diode_orientation = DiodeOrientation.COLUMNS

keyboard.bluefruitspi_rst = board.D11
keyboard.bluefruitspi_irq = board.D12
keyboard.bluefruitspi_cs = board.D13

keyboard.keymap = [
    [
        KC.A, KC.B,
        KC.LCTRL, KC.D,
    ]
]

if __name__ == '__main__':
    keyboard.go(hid_type=HIDModes.BLUEFRUITSPI)
```

This specific implementation is meant to minimize the time spent moving data around, and is hence somewhat specific to `AT+BLEKEYBOARDCODE`, but may be worth generalizing to other commands as well.

### CODE REVIEW

**Comment by ladyada** [comment]
nifty!  was this tested on a feather m0 bluefruit? if not, which was it tested on?

**Comment by Gigahawk** [comment]
Ah sorry I should have mentioned, tested on a Feather STM32F405 Express, although the KMK example should work on any Circuitpython compatible board if you just update the pin definitions.

**Comment by ladyada** [comment]
it needs to be tested on a feather m0 BLE to verify that there's no memory problems
@hathach @dhalbert do you have one to test with?

**Comment by ladyada** [comment]
@Gigahawk there are CI failures, please check em out and fix the warnings/errors! https://github.com/adafruit/Adafruit_CircuitPython_BluefruitSPI/pull/20/checks?check_run_id=2714893932
here's a guide on how to do it
https://learn.adafruit.com/contribute-to-circuitpython-with-git-and-github

**Comment by Gigahawk** [comment]
Yea my bad I just noticed I missed the FIFO instantiation.

I don't have an m0 to test this on, but the additional memory requirement for this should be fairly minimal right?
For any memory constrained device you would just use a shorter buffer length.

**Comment by ladyada** [comment]
hopefully :D

File: adafruit_bluefruitspi.py [hunk 1]
```diff
@@ -166,12 +178,19 @@ def send_keyboard_code(self, evt):
             self._fifo_buffer.enqueue(k)
 
     def pop_keyboard_code_queue(self):
+        """Send an SDEP packet from the FIFO buffer to the Bluefruit"""
         data = self._fifo_buffer.dequeue()
         if data is not None:
```
Comment (Gigahawk) [inline adafruit_bluefruitspi.py:183]:
```
************* Module adafruit_bluefruitspi
adafruit_bluefruitspi.py:183:4: R0201: Method could be a function (no-self-use)
```
I'm not sure what the issue is here, this line isn't even pointing to a method definition, and the method it's within definitely refers to `self`

https://github.com/Gigahawk/Adafruit_CircuitPython_BluefruitSPI/runs/2722629440?check_suite_focus=true

Comment (ladyada) [reply adafruit_bluefruitspi.py:183]: hmm not sure either.. @dhalbert any idea

Comment (tannewt) [reply adafruit_bluefruitspi.py:183]: `black` also ran and shifted the line numbering a bit. I think its referring to `_create_sdep_raw` which doesn't use self. You can remove self and then mark the function with `@staticmethod`.

Comment (Gigahawk) [reply adafruit_bluefruitspi.py:183]: Thanks! That was the problem https://github.com/Gigahawk/Adafruit_CircuitPython_BluefruitSPI/runs/2731338134?check_suite_focus=true

**Comment by Gigahawk** [comment]
Is what this PR proposes OK?

I've been thinking about how to generalize this for other commands.
It would be impractical to preallocate SDEP packets for every single command that might benefit from a FIFO like this.
Instead I was thinking of moving the allocation into `init_keycode_template()`, and making it public with the expectation that users would call it prior to using any of the other dependent commands.
I suppose the FIFO buffer instantiation could also be moved to a separate function for maximum memory savings.

**Comment by ladyada** [comment]
we need to have it tested on an m0 bluefruit before it's merged!

**Comment by Gigahawk** [comment]
What I have implemented right now is mostly a proof of concept.

What I'm asking is how you would like the API to work, keeping in mind that this feature may be memory intensive and should probably be off by default.

**Comment by FoamyGuy** [comment]
@Gigahawk Are you still interested in working on this one? It needs merged from main / conflicts resolved atm. I do think the proof of concept that you've implemented here seems like a good start. Further down the road it could be nice to generalize it for other commands as well, but cool to start with this one and get it merged once ready and tested.

**Comment by Gigahawk** [comment]
@FoamyGuy merge conflict fixed, although I'm no longer actively using/testing this code since QMK mostly supports the BluefruitSPI with non AVR processors now.

File: adafruit_bluefruitspi.py [hunk 2]
```diff
@@ -30,6 +30,7 @@
 
 import time
 import struct
+import binascii
```
Comment (jerryneedell) [inline adafruit_bluefruitspi.py]: on an M0 board -- like the feather_m0_bluefruit (using the feather_m0_adalogger build), the binascii module is not compiled in so it has to be imported from the adafruit_binascii.mpy module

the simplest fails at this point on the feather_m0_bluefruit (adalogger) build.

Comment (jerryneedell) [reply adafruit_bluefruitspi.py]: I modified the code to import adafruit_binascii and  the simpletest runs OK on the feather_m0_bluefruitt with the feather_m0_adalogger build
```
Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:

Code done running.

Press any key to enter the REPL. Use CTRL-D to reload.

Adafruit CircuitPython 7.2.0-alpha.1-203-gbf93ddec5 on 2022-01-24; Adafruit Feather M0 Adalogger with samd21g18
>>> import bluefruitspi_simpletest
Initializing the Bluefruit LE SPI Friend module
BLESPIFRIEND
nRF51822 QFACA10
78585487A8785672
0.8.1
0.8.1
Apr 10 2019
S110 8.0.0, 0.2

Waiting for a connection to Bluefruit LE Connect ...
...................................................
 *Connected!*
```

Perhaps a try/except can be used to select the proper import

**Review by jerryneedell — CHANGES_REQUESTED** [review]

**Comment by jerryneedell** [comment]
making this change seemed to work ok on the feather_m0_bluefruit
```
try: 
    import binascii
except:
    import adafruit_binascii
```

**Comment by jerryneedell** [comment]
Note: I did not review the code in detail or test any functionality beyond a UART transfer with the "simpletest" and the BLE Connect App on my iPhone.

**Review by jerryneedell — COMMENTED** [review]
Retested on feather_m0_bluefruit - bluefruitspi_simpletest.py from the examples loads and runs. I was able to connect and transfer data via the UART function of the BLE Connect App.
The import of binascii has been corrected.
This test had been run on the 7.2.0-alpha.1 latest code from main.
I also verified that the bluefruitspi_simpletest loads and runs normally on the stable release 7.1.1

**Review by jerryneedell — APPROVED** [review]

**Comment by jerryneedell** [comment]
Note: My approval is meant to confirm that the library functions as expected on the feather_m0_bluefruit.
I cannot comment on the functionality of the new features added by this PR.

**Review by ladyada — APPROVED** [review]

**Comment by ladyada** [comment]
sounds good to us - note not many folks use this library (since nrf52840 has native ble!)

**Pull request merged**